Additional egs++ geometries and shapes

randlet opened the pull request:
As discussed previously this PR introduces a number of new egs++ geometries and shapes outined below (example geometry files for the new geometries are also included).

One thing of note is that there are a couple of other GPL'd files included (gzstream.cpp, gzstream.h & sobol.cpp, sobol.h) that I did not write. Hopefully that doesn't cause any issues for distribution.

Hopefully everything is in OK shape but I recognize that this is a big PR so we may have to go through a few revisions (e.g. do you want this all squashed to a single commit?).

If you need a more formal statement of transfer of copyright than those contained in commit messages then please let me know!
## Geometry
### EGS_AutoEnvelope

An envelope geometry that inscribes a single geometry at one or more locations within a base geometry (think brachy seeds in a phantom). The primary difference between the regular egs++ geometry envelope and the auto-envelope is that, at initialization time the auto-envelope uses a Monte-Carlo routine to discover which regions of the base geometry are occupied by the inscribed geometry.  This allows us to skip bounds checking of the inscribed geometries when in regions of the base geometry that have no inscribed geometries in them.  This makes the transport in the phantom (roughly) independent of the number of inscribed geometries and can give a large gain in efficiency for some simulations.  This library also includes an additional EGS_ASwitchedEnvelope that allows user codes to enable / disable inscribed geometries arbitrarily (we use this for simulating inter-seed effects in TG43 calculations and stepped HDR sources).
### EGS_rz

A wrapper around ND_geometry to simplify the creation of RZ geometries.
### EGS_cSpheres

We've also modified the EGS_cSpheres geometry class to a) implement  getBound, getNRegDir,  and getMass and b) implement a spherical shell class (innermost sphere is considered to be outside the geometry). 
### EGS_glib

A very small shim library that allows you to load geometries from external files (similar to an "include" directive) like so:

:start geometry:
    library = egs_glib
    name = my_external_geom
    include file = /path/to/some/external/my_geom.geom
:stop geometry:

Where the external file is a valid `geometry definition` input block.  The advantages of this over a regular "include" is that you can now use egs_view to view the geometry in your external file and use "my_external_geom" in other composite geometries in your input file. This geometry can also be used to create an XYZ geometry from an egsphant file.
## Shapes
### egs_conical_shell_stack

Allows you to define a shape in a similar way to the egs cone stack geometry.
### egs_spherical_shell

Allows you to create a thin spherical shell shape.  This shell can also be truncated by a conical section of a given half-angle (we use it for simulating beta-emitting eye-plaques).

ftessier commented:
Great work @randlet, these are all extremely useful addition to the egs++ library! We cannot include in the project anything over which NRC does not own copyright. How critical are gzstream and sobol to this PR?

We **don't** want to squash larger PR such as this one! We want to preserve your commit history and add a proper merge commit to develop. Typically we only squash smaller changes, so as to limit the number of small merge bubbles.

We will probably take some time to review, but I am tagging it for the next release.

randlet commented:
The gzstream and Sobol RNG code is for efficiency (disk space in the case of gzstream and cpu time in the case of Sobol) and not crucial to any of the libraries.  Perhaps I could make use of preprocessor directives and allow people to conditionally compile those features after they download the gzstream & Sobol RNG themselves?

ftessier commented:
Sounds like a good idea: leave it as an option and provide instructions to
load the third party libraries. We have some sobol code, so perhaps we can
use that.
On Wed, Jun 29, 2016 at 08:07 Randle Taylor <notifications@github.com>
wrote:

> The gzstream and Sobol RNG code is for efficiency (disk space in the case
> of gzstream and cpu time in the case of Sobol) and not crucial to any of
> the libraries. Perhaps I could make use of preprocessor directives and
> allow people to conditionally compile those features after they download
> the gzstream & Sobol RNG themselves?
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nrc-cnrc/EGSnrc/pull/103#issuecomment-229336940>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ABYJkVu7j748SmFvBJxcPtWr7D5Gkxurks5qQl_vgaJpZM4IxmnG>
> .
>

randlet commented:
I've updated this so that everything will now compile and run without the gzstream & Sobol rng dependencies.  Users who want the extra functionality can obtain the code required from [here](https://github.com/clrp-code/egspp-geometry-lib-extras).

mchamberland commented:
@randlet if a user wants to install the gzip functionality on Mac, the egs_autoenvelope and egs_glib Makefiles need to explicitly link against the 'z' library to compile. So in your repository for extra functionality, you should include modified Makefiles for auto envelope and glib.

egs_autoenvelope:
[Line 35](https://github.com/nrc-cnrc/EGSnrc/pull/103/files#diff-961951a6b1c090ace1083f1d6c94c298R35) should be:
`link2_libs = egs_gtransformed egspp z`

egs_glib:
[Line 25](https://github.com/nrc-cnrc/EGSnrc/pull/103/files#diff-9e9acc28a4794cc558e282ea000a823aR25) should be:
`link2_libs = egspp egs_ndgeometry egs_planes z`

mchamberland commented:
Or maybe a better way is having the following in the original Makefiles (using the egs_autoenvelope Makefile as an example):

```
GZSTREAM =
GZSTREAM_DEF =
GZSTREAM_H =
GZSTREAM_LIB =
ifneq ("$(wildcard gzstream.cpp)","")
GZSTREAM = gzstream
GZSTREAM_DEF = -DBUILD_GZSTREAM -DHAS_GZSTREAM
GZSTREAM_H = gzstream.h
GZSTREAM_LIB = z
endif

DEFS = $(DEF1) -DBUILD_AENVELOPE_DLL $(SOBOL_DEF) $(GZSTREAM_DEF)


library = egs_autoenvelope
lib_files = egs_autoenvelope $(SOBOL) $(GZSTREAM)

link2_libs = egs_gtransformed egspp $(GZSTREAM_LIB)

```

mchamberland commented:
@randlet Also, I think [line 43](https://github.com/nrc-cnrc/EGSnrc/pull/103/files#diff-961951a6b1c090ace1083f1d6c94c298R43) in egs_autoenvelope/Makefile should be:

$(DSO2)autoenvelope.$(obje): volcor.h $(GZSTREAM_H) \

so, remove the $(SOBOL_H) part. With the $(SOBOL_H) part, the sobol.o file does not get created under the egs++/dso folder, which leads to compilation errors elsewhere. Removing the $(SOBOL_H) seems to resolve this.

randlet commented:
Thanks @mchamberland  :)

rtownson commented on HEN_HOUSE/egs++/geometry/egs_rz/rz1.geom:
```diff
@@ -0,0 +1,18 @@
+
+
+:stop geometry definition:
+
```
This file is not needed - there is already an example in `examples/rz_lib1.geom`.

rtownson commented on HEN_HOUSE/egs++/geometry/egs_rz/rz2.geom:
```diff
@@ -0,0 +1,22 @@
+    simulation geometry = phantom
+
+:stop geometry definition:
+
```
This file is not needed - there is already an example in `examples/rz_lib2.geom`.

rtownson commented on HEN_HOUSE/egs++/geometry/egs_rz/egs_rz.cpp:
```diff
@@ -0,0 +1,248 @@
+*/
+
+
+/*! \file egs_rz.h
```
Should be `/*! \file egs_rz.cpp`

rtownson commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/egs_autoenvelope.h:
```diff
@@ -0,0 +1,589 @@
+Geometry Definition Format
+--------------------------
+
+There is an example geometry included in the egs_autoenvelope/6702.geom file.
```
I think you mean examples/seeds_in_xyz_aenv.geom

rtownson commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/egs_autoenvelope.cpp:
```diff
@@ -0,0 +1,1018 @@
+        egsFatal("EGS_AEnvelope:: Failed to find any regions with inscribed geometries\n");
+    }
+
+    if (output_vc=="text" || output_vc=="gzip") {
```
We're also supposed to be able to use "yes" as equivalent to "text", so this line should be:

`if (output_vc=="yes" || output_vc=="text" || output_vc=="gzip") {`

rtownson commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/egs_autoenvelope.h:
```diff
@@ -0,0 +1,589 @@
+
+        inscribed geometry name = seed # the name of a previously defined geometry
+
+        output volume correction file = text # (optional) text or gzip
```
I think that using `output volume correction file` here does nothing, so remove it from the example.

rtownson commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/egs_autoenvelope.cpp:
```diff
@@ -0,0 +1,1018 @@
+        }
+    }
+
+    vc_results.outputResults();
```
I get a segmentation fault around here if you load volumes from a volume correction file (used "text" mode). I think we'll need to do some robustness testing around the `volume correction file` option. If you comment this line out, there is no segmentation fault.

randlet commented:
Thanks for all the review @rtownson!  Can you ping me when you've finished your initial review and I'll get everything updated.

rtownson commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/volcor.h:
```diff
@@ -0,0 +1,663 @@
+}
+
+int loadVolumes(string fname, vector<RegVolume> &reg_volumes, RegionGeomSetT &reg_with_inscribed, vector<EGS_BaseGeometry *> inscribed) {
+    ifstream vfile(fname.c_str(), ios::binary);
```
This fails for "text" format volume files, because they're not binary. Try something like:

```
int loadVolumes(string fname, vector<RegVolume> &reg_volumes, RegionGeomSetT &reg_with_inscribed, vector<EGS_BaseGeometry *> inscribed) {
    ifstream vfile(fname.c_str(), ios::binary);
    if (!vfile.is_open()) {
        return 1;
    }
    
    if (isGZip(vfile)) {
        vfile.close();
#ifdef HAS_GZSTREAM
        igzstream gzf(fname.c_str());
        readVolumes(gzf, reg_volumes, reg_with_inscribed, inscribed);
        gzf.close();
#else
        egsWarning("Tried to load gzip volume correction file but not compiled with gzstream.\n");
        return 1;
#endif
    }
    else {
        vfile.close();
        ifstream vfile2(fname.c_str(), ios::in);
        readVolumes(vfile2, reg_volumes, reg_with_inscribed, inscribed);
        vfile2.close();
    }

    return 0;

}
```

rtownson commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/volcor.h:
```diff
@@ -0,0 +1,663 @@
+vector<EGS_Float> getUncorrectedVolumes(EGS_BaseGeometry *base) {
+    vector<EGS_Float> uncorrected;
+    for (int ir=0; ir < base->regions(); ir++) {
+        EGS_Float vol = base->getMass(ir)/base->getRelativeRho(ir);
```
Not all geometries have `getMass`! For example, `egs_box`.

What happens if we can't calculate the mass for the base geometry, because it is complex? Maybe there should be a check and return an error for geometries that are not supported.

rtownson commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/volcor.h:
```diff
@@ -0,0 +1,663 @@
+    EGS_AffineTransform *first_transform = transforms[0];
+
+    HitCounterT hit_counter;
+
```
Incrementing the hit counter without initializing it makes me nervous (though it seems to work). Add something like:

```
    int nreg = base->regions();
    for (int i=0; i < nreg; ++i) {
        hit_counter[i] = 0;
    }
```

randlet replied:
@rtownson I looked into this when implementing and iirc its in the standard that std::map gets default initialized (http://stackoverflow.com/questions/4523959/stdmap-default-value-for-build-in-type backs this up) which means the value for any key will be zeroed (0) in this case.  Initializing every possible region would waste memory since generally in e.g. brachy applications we will have a large number of regions, most of which won't have sources in them.  

Furthermore, other code relies on hit_counter only having keys for regions actually containing geometries so if we explicitly initialize hit counter for every region, we would need to implement more bookkeeping. 

Hope that makes sense!

rtownson replied:
Sounds good to me! We can revisit this section of the code when we look at integrating the other volume calculation algorithm (Fred is working on improving it further first, and there is a bug in the current PR).

rtownson commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/volcor.h:
```diff
@@ -0,0 +1,663 @@
+ * the volume of those regions. The algorithm is described
+ * on main page of the Auto Envelope documentation. */
+VCResults findRegionsWithInscribed(VCOptions *opts, EGS_BaseGeometry *base,
+                                   vector<EGS_BaseGeometry *> inscribed, vector<EGS_AffineTransform *> transforms) {
```
We should look into making use of the more efficient volume calculations in pull request #184.

rtownson commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/egs_autoenvelope.h:
```diff
@@ -0,0 +1,589 @@
+*
+*  The inputs are all exactly the same as the EGS_AEnvelope and only the
+*  first inscribed geometry is activated upon initialization. Examples of using
+*  this geometry type are available in the EGS_ASwitchedEnvlope class documentation.
```
Can remove this last sentence since we're currently in the class documentation.

rtownson commented on HEN_HOUSE/egs++/geometry/egs_glib/egs_glib.cpp:
```diff
@@ -0,0 +1,290 @@
+    return med_rhos;
+}
+
+EGS_BaseGeometry *readEGSPhant(istream &data, map<string, EGS_Float> med_rhos) {
```
Egsphant files are also read by `egs_nd_geometry`, perhaps we should call this function from there to avoid code duplication.

randlet replied:
Ahhh, thanks for pointing this out, I completely missed that we could already do this in ND geometry.  Yes, there's no sense in duplicating this functionality, and it's done more completely in ND geometry so I'll strip it out of glib.

rtownson commented on HEN_HOUSE/egs++/geometry/egs_rz/egs_rz.cpp:
```diff
@@ -0,0 +1,248 @@
+
+        rz->setName(input);
+        rz->setMedia(input);
+
```
Need to set labels also:

`rz->setLabels(input);`

rtownson commented on HEN_HOUSE/egs++/shapes/egs_spherical_shell/egs_spherical_shell.cpp:
```diff
@@ -0,0 +1,150 @@
+        err = input->getInput("hemisphere", hemisphere);
+        if (err) {
+            hemisphere = 0;
+        }
```
Add more thorough input validation. Inner radius > outer radius, negative radii, hemisphere != -1 or 1 should all throw errors.

rtownson commented on HEN_HOUSE/egs++/shapes/egs_spherical_shell/egs_spherical_shell.h:
```diff
@@ -0,0 +1,133 @@
+\verbatim
+:start shape:
+    library = egs_spherical_shell
+    midpoint = O, O, O
```
These O's should be 0's for happy copy/pasting :)

rtownson commented on HEN_HOUSE/egs++/shapes/egs_conical_shell/egs_conical_shell.h:
```diff
@@ -0,0 +1,245 @@
+
+public:
+
+    /*! \brief Construct a sphere of radius \a r with midpoint \a Xo */
```
Should be:
`Construct a conical shell stack with midpoint \a Xo`

rtownson commented on HEN_HOUSE/egs++/shapes/egs_conical_shell/egs_conical_shell.cpp:
```diff
@@ -0,0 +1,279 @@
+            ++nl;
+        }
+
+        /*    ofstream out;
```
Can remove this commented out code.

randlet commented:
Many thanks for the careful review @rtownson!  I will get the fixes made and take a look at switching to the new algorithm asap.

randlet commented:
@rtownson I believe I've addressed all the issues you pointed out now.  I stripped out the egsphant code from egs_glib and added optional support for [using compressed egsphant files to egs_ndgeometry](https://github.com/nrc-cnrc/EGSnrc/pull/103/commits/27803530e39ead7bc2048cacd173f91cbbafc2e8) so you should take a look at that and make sure you're ok with it.

randlet commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/egs_autoenvelope.cpp:
```diff
@@ -0,0 +1,1046 @@
+
+    egsInformation(
+        "Writing volume correction file for %s to %s file %s\n",
+        gname.c_str(), output_vc.c_str(), fname.c_str());
```
Fixed!

ftessier commented:
Given the size of this commit, we ought to merge into develop after the 2017 release and tag for 2018.

rtownson commented:
I think we can merge this with develop now.

ftessier commented:
Agreed, but we ought to sort out the copyright transfer; this is a significant addition to the egs++ library!

mchamberland commented:
Just curious, have you been in touch with @randlet or Dave and Rowan at Carleton about the copyright transfer yet? I can ping them about it if needed.

ftessier commented:
Thanks. Not yet, I am waiting on NRC legal to draft a single copyright transfer agreement for external EGSnrc contributions.

ftessier commented:
@randlet I will send you a draft copyright transfer agreement shortly, for review. In the mean time, please:

1. format your pull request through the [EGSnrc astyle standard](https://github.com/nrc-cnrc/EGSnrc/wiki/Coding-style): `cd $HEN_HOUSE/egs++; astyle --recursive --options=$HEN_HOUSE/scripts/egs_style_standard "*.h" "*.cpp"`

2. Adjust the authorship part of the header. Author format should be a single <firstname lastname, year>. Contributors list additional people (format <firstname lastname>) who worked on the file, even for small modifications. This header section contains only names, not institutions.

3. If you want to add more information, for example say that you are submitting on behalf of CLRP etc., add another header section below the authorship, similar to [beamnrc.mortran](https://github.com/nrc-cnrc/EGSnrc/blob/master/HEN_HOUSE/omega/beamnrc/beamnrc.mortran#L42-L71) for example.

4. We will also add a header section to describe and date the copyright transfer agreement. I can do that at the end.

ftessier commented:
@randlet, this PR fails to compile on Travis, perhaps rebasing on the current develop head would solve the issue? As an aside, we had a meeting with legal this morning to discuss the copyright assignment. They understand the issues we raised, and they will provide a cleaner draft as soon as possible.

ftessier commented:
@randlet we ought to remove files `gzstream.cpp`, `gzstream.h`, `sobol.cpp` and `sobol.h` from this branch. I know they are removed at the branch tip, but since I plan on merging this, the files will be distributed in the repository, namely in commit 09fab65 is one checked out this commit. Can you `rebase -i` and edit this commit to remove those files from the commit history completely? Otherwise I can take care of it before merging.

randlet commented:
@ftessier I fixed an issue with the Makefile's for a couple of the geometries and rebased on current develop, removing the 3rd party files.  Build is now passing and I think history should be clean of those files now.

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/egs_autoenvelope.cpp:
```diff
@@ -0,0 +1,1057 @@
+###############################################################################
+#
+#  EGSnrc egs++ auto envelope geometry
+#  Copyright (C) 2015 National Research Council Canada
```
```suggestion
#  Copyright (C) 2015 Randle E.P. Taylor, Rowan M. Thomson, Marc J.P. Chamberland, and Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/egs_autoenvelope.cpp:
```diff
@@ -0,0 +1,1057 @@
+#
+#  Author:          Randle Taylor, 2016
+#
+#  Contributors:
```
```suggestion
#  Contributors: Marc J.P. Chamberland
                 Rowan M. Thomson
                 Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/egs_autoenvelope.h:
```diff
@@ -0,0 +1,600 @@
+###############################################################################
+#
+#  EGSnrc egs++ auto envelope geometry headers
+#  Copyright (C) 2015 National Research Council Canada
```
```suggestion
#  Copyright (C) 2015 Randle E.P. Taylor, Rowan M. Thomson, Marc J.P. Chamberland, and Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/egs_autoenvelope.h:
```diff
@@ -0,0 +1,600 @@
+#
+#  Author:          Randle Taylor, 2016
+#
+#  Contributors:
```
```suggestion
#  Contributors: Marc J.P. Chamberland
                 Rowan M. Thomson
                 Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/volcor.h:
```diff
@@ -0,0 +1,708 @@
+###############################################################################
+#
+#  EGSnrc egs++ envelope geometry headers
+#  Copyright (C) 2015 National Research Council Canada
```
```suggestion
#  Copyright (C) 2015 Randle E.P. Taylor, Rowan M. Thomson, Marc J.P. Chamberland, and Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_autoenvelope/volcor.h:
```diff
@@ -0,0 +1,708 @@
+#
+#  Author:          Randle Taylor, 2016
+#
+#  Contributors:
```
```suggestion
#  Contributors: Marc J.P. Chamberland
                 Rowan M. Thomson
                 Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_glib/egs_glib.cpp:
```diff
@@ -0,0 +1,83 @@
+###############################################################################
+#
+#  EGSnrc egs++ glib geometry
+#  Copyright (C) 2015 National Research Council Canada
```
```suggestion
#  Copyright (C) 2015 Randle E.P. Taylor, Rowan M. Thomson, Marc J.P. Chamberland, and Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_glib/egs_glib.cpp:
```diff
@@ -0,0 +1,83 @@
+#
+#  Author:          Randle Taylor, 2016
+#
+#  Contributors:
```
```suggestion
#  Contributors: Marc J.P. Chamberland
                 Rowan M. Thomson
                 Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_glib/egs_glib.h:
```diff
@@ -0,0 +1,189 @@
+###############################################################################
+#
+#  EGSnrc egs++ glib geometry headers
+#  Copyright (C) 2015 National Research Council Canada
```
```suggestion
#  Copyright (C) 2015 Randle E.P. Taylor, Rowan M. Thomson, Marc J.P. Chamberland, and Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_glib/egs_glib.h:
```diff
@@ -0,0 +1,189 @@
+#
+#  Author:          Randle Taylor, 2016
+#
+#  Contributors:
```
```suggestion
#  Contributors: Marc J.P. Chamberland
                 Rowan M. Thomson
                 Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_rz/egs_rz.cpp:
```diff
@@ -0,0 +1,275 @@
+###############################################################################
+#
+#  EGSnrc egs++ rz geometry
+#  Copyright (C) 2015 National Research Council Canada
```
```suggestion
#  Copyright (C) 2015 Randle E.P. Taylor, Rowan M. Thomson, Marc J.P. Chamberland, and Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_rz/egs_rz.cpp:
```diff
@@ -0,0 +1,275 @@
+#
+#  Author:          Randle Taylor, 2016
+#
+#  Contributors:
```
```suggestion
#  Contributors: Marc J.P. Chamberland
                 Rowan M. Thomson
                 Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_rz/egs_rz.h:
```diff
@@ -0,0 +1,161 @@
+###############################################################################
+#
+#  EGSnrc egs++ rz geometry headers
+#  Copyright (C) 2015 National Research Council Canada
```
```suggestion
#  Copyright (C) 2015 Randle E.P. Taylor, Rowan M. Thomson, Marc J.P. Chamberland, and Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/geometry/egs_rz/egs_rz.h:
```diff
@@ -0,0 +1,161 @@
+#
+#  Author:          Randle Taylor, 2016
+#
+#  Contributors:
```
```suggestion
#  Contributors: Marc J.P. Chamberland
                 Rowan M. Thomson
                 Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/shapes/egs_conical_shell/egs_conical_shell.cpp:
```diff
@@ -0,0 +1,272 @@
+###############################################################################
+#
+#  EGSnrc egs++ conical shell shape
+#  Copyright (C) 2015 National Research Council Canada
```
```suggestion
#  Copyright (C) 2015 Randle E.P. Taylor, Rowan M. Thomson, Marc J.P. Chamberland, and Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/shapes/egs_conical_shell/egs_conical_shell.cpp:
```diff
@@ -0,0 +1,272 @@
+#
+#  Author:          Randle Taylor, 2016
+#
+#  Contributors:
```
```suggestion
#  Contributors: Marc J.P. Chamberland
                 Rowan M. Thomson
                 Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/shapes/egs_conical_shell/egs_conical_shell.h:
```diff
@@ -0,0 +1,249 @@
+###############################################################################
+#
+#  EGSnrc egs++ conical shell headers
+#  Copyright (C) 2015 National Research Council Canada
```
```suggestion
#  Copyright (C) 2015 Randle E.P. Taylor, Rowan M. Thomson, Marc J.P. Chamberland, and Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/shapes/egs_conical_shell/egs_conical_shell.h:
```diff
@@ -0,0 +1,249 @@
+#
+#  Author:          Randle Taylor, 2016
+#
+#  Contributors:
```
```suggestion
#  Contributors: Marc J.P. Chamberland
                 Rowan M. Thomson
                 Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/shapes/egs_spherical_shell/egs_spherical_shell.cpp:
```diff
@@ -0,0 +1,200 @@
+#
+#  Author:          Randle Taylor, 2016
+#
+#  Contributors:
```
```suggestion
#  Contributors: Marc J.P. Chamberland
                 Rowan M. Thomson
                 Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/shapes/egs_spherical_shell/egs_spherical_shell.cpp:
```diff
@@ -0,0 +1,200 @@
+###############################################################################
+#
+#  EGSnrc egs++ spherical shell shape
+#  Copyright (C) 2015 National Research Council Canada
```
```suggestion
#  Copyright (C) 2015 Randle E.P. Taylor, Rowan M. Thomson, Marc J.P. Chamberland, and Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/shapes/egs_spherical_shell/egs_spherical_shell.h:
```diff
@@ -0,0 +1,137 @@
+###############################################################################
+#
+#  EGSnrc egs++ spherical shell shape headers
+#  Copyright (C) 2015 National Research Council Canada
```
```suggestion
#  Copyright (C) 2015 Randle E.P. Taylor, Rowan M. Thomson, Marc J.P. Chamberland, and Dave W.O. Rogers
```

mchamberland commented on HEN_HOUSE/egs++/shapes/egs_spherical_shell/egs_spherical_shell.h:
```diff
@@ -0,0 +1,137 @@
+#
+#  Author:          Randle Taylor, 2016
+#
+#  Contributors:
```
```suggestion
#  Contributors: Marc J.P. Chamberland
                 Rowan M. Thomson
                 Dave W.O. Rogers
```

ftessier commented:
@randlet can you open your branch to [allow edits from maintainers](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)? I would like to polish up the branch before merging into develop. Alternatively, I can open a new pull request, but I would prefer to keep it all tied to #103.

ftessier commented:
@marenaud we typically don't include middle initials for the Author and Contributors in the file headers, ok to remove? However we can leave them in for the copyright line. Waiting to hear from @randlet for allowing edits by maintainers to start committing your changes.

marenaud commented:
@ftessier I think you @'d the wrong Marc ;) @mchamberland

ftessier commented:
Yes I did, thanks! :confounded:

mchamberland commented:
@ftessier  Okay with me to remove in the Author and Contributors.

randlet commented:
@ftessier  I enabled that option. Thanks!

ftessier commented:
@randlet **thank you!** I am cleaning up the branch locally and will `push --force-with-lease` later this week. In the mean time, please commit any changes on **top** of this branch (i.e., don't change the branch history so far) so I can easily replay additional commits on top of my pared down version.

ftessier commented:
Rebased on `develop`. Re-ordered commits and squashed some. No changes to the code as confirmed by a null diff with the original branch. Some additional changes to come. In the mean time please add any changes on top of this branch, i.e., don't force-push, so that I can easily integrate interim modifications in the revised history.

ftessier commented:
Reorganized and squashed a little further. Still no diff with the original `randlet/feature-geometries`.

ftessier commented:
No changes still. I had just messed up authorship while fixing merge conflicts with gitkraken.

ftessier commented:
This does not compile since it was rebased on `develop`, because of pull request #364 after this pull request was created, and changed the way the projectors are called.

rtownson commented:
EGS_AutoEnvelope is still non-functional (memory leak).

ftessier commented:
Fixed the ZProjector compilation error, and tweaked typos and format slightly.

ftessier commented:
Condensed into 9 "orthogonal" commits: they now all pertain to a single directory. Squashed style and comments commits. Separated out the glib from the auto envelope commits (which made me the commit author unfortunately). Ported the small commit to curb the compiler warning for `egs_cylinders.cpp` to the `small-stuff` branch. The diff with the original branch does not otherwise show any functional code changes. Commit messages are all messed up from all the squashing and reordering; will update them now.

ftessier commented:
Updated the commit messages according to the new commit line. Also updated the copyright and contributors for the new sample input files to reflect the ones found in the corresponding source code additions. Verified the diff with the original branch.

ftessier commented:
This branch compiles without errors, and I checked that I can run an autoenvelope simulation, and load in egs_view the example `HEN_HOUSE/egs++/geometry/examples/seeds_in_xyz_aenv.geom`. Hence I am merging this into `develop`.

ftessier commented:
Rebased on `develop`.

ftessier commented:
@rtownson is the segmentation fault resolved at your end?

rtownson commented:
@ftessier Yes, it's possible that the rebase fixed the segmentation fault issue. I'm ok with merging this in.
